source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,528 @@
+    connection->detectEarlyCloseWhenReadDisabled(
+        per_connection_state_.detect_early_close_when_read_disabled_.value());
+  }
+  if (per_connection_state_.no_delay_.has_value()) {
+    connection->noDelay(per_connection_state_.no_delay_.value());
+  }
+  if (per_connection_state_.connection_stats_.has_value()) {
+    connection->setConnectionStats(*per_connection_state_.connection_stats_);
+  }
+  if (per_connection_state_.buffer_limits_.has_value()) {
+    connection->setBufferLimits(per_connection_state_.buffer_limits_.value());
+  }
+  if (per_connection_state_.enable_half_close_.has_value()) {
+    connection->enableHalfClose(per_connection_state_.enable_half_close_.value());
+  }
+  if (per_connection_state_.delayed_close_timeout_.has_value()) {
+    connection->setDelayedCloseTimeout(per_connection_state_.delayed_close_timeout_.value());
+  }
+  if (per_connection_state_.start_secure_transport_.has_value()) {
+    ASSERT(per_connection_state_.start_secure_transport_);
+    connection->startSecureTransport();
+  }
+
+  return connection;
+}
+
+void HappyEyeballsConnectionImpl::tryAnotherConnection() {
+  connections_.push_back(createNextConnection());
+  connections_.back()->connect();
+  maybeScheduleNextAttempt();
+}
+
+void HappyEyeballsConnectionImpl::maybeScheduleNextAttempt() {
+  if (next_address_ >= address_list_.size()) {
+    return;
+  }
+  next_attempt_timer_->enableTimer(std::chrono::milliseconds(300));
+}
+
+void HappyEyeballsConnectionImpl::onEvent(ConnectionEvent event,
+                                          ConnectionCallbacksWrapper* wrapper) {
+  wrapper->connection().removeConnectionCallbacks(*wrapper);
+  if (event != ConnectionEvent::Connected) {
+    if (next_address_ < address_list_.size()) {
+      next_attempt_timer_->disableTimer();
+      tryAnotherConnection();
+    }
+    if (connections_.size() > 1) {
+      // Nuke this connection and associated callbacks and let a subsequent attempt proceed.
+      cleanupWrapperAndConnection(wrapper);
+      return;
+    }
+  }
+
+  connect_finished_ = true;
+  next_attempt_timer_->disableTimer();
+
+  // Close and delete up other connections.
+  auto it = connections_.begin();
+  while (it != connections_.end()) {
+    if (it->get() != &(wrapper->connection())) {
+      (*it)->removeConnectionCallbacks(*wrapper);
+      (*it)->close(ConnectionCloseType::NoFlush);
+      it = connections_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+  ASSERT(connections_.size() == 1);
+  callbacks_wrappers_.clear();
+
+  // Apply post-connect state to the final socket.
+  for (auto cb : post_connect_state_.connection_callbacks_) {
+    if (cb) {
+      connections_[0]->addConnectionCallbacks(*cb);
+    }
+  }
+
+  for (auto cb : post_connect_state_.bytes_sent_callbacks_) {
+    connections_[0]->addBytesSentCallback(cb);
+  }
+
+  if (event == ConnectionEvent::Connected) {
+    for (auto& filter : post_connect_state_.filters_) {
```

Comment:
optional: helper function?
less optional: comment this is doing defered actions which shouldn't be applied to the connections not in use

Reply:
> less optional: comment this is doing defered actions which shouldn't be applied to the connections not in use

Hm. That's what the comment on line 449 is attempt to convey. (Starting from there is where we start applying state.) The state is this block is only applied if the final connection succeeds. (It might be the case that the final connection attempt actually fails) I suppose we could also apply this state even when the final connection fails. Do you think that would be preferable?

> optional: helper function?

Would the helper function be for applying state to the final connection? (Like everything from line 449 on down?)